### PR TITLE
Add frazil to ice-shelf melt rates

### PR DESCRIPTION
**Pull request by irenavankova**
New variable frazilIceFreshwaterFlux is introduced to keep track of frazil ice formation instanatenously (and not only commulatively over coupling timestep as before)

New variable landIceFreshwaterFluxTotal is introduced to keep track of total melt rates that includes both interface melt/freeze and frazil ice formation, so that landIceFreshwaterFluxTotal = landIceFreshwaterFlux + frazilIceFreshwaterFlux.

Associted changes in MPAS-Analysis will reflect this to plot Antarctic meltrate maps with landIceFreshwaterFluxTotal instead of landIceFreshwaterFlux.

The changes here affect diagnostics only.

The plot here shows the monthly averaged fields (landIceFreshwaterFlux, frazilIceFreshwaterFlux, and landIceFreshwaterFluxTotal) for the year 6 month 12 of a Gcase with the ECwISC30to60E2r1 mesh. Freezing patterns over the Filchner-Ronne look much more realistic with landIceFreshwaterFluxTotal. 

Screen Shot 2024-02-25 at 10 23 01 PM

### CODE REVIEW

**Comment by xylar** [comment]
@irenavankova, also thanks for the figures above!  I'm so glad this work turned out to be important!  I wouldn't have wanted you to put this time in if it turned out to be a negligible effect.

File: components/mpas-ocean/cime_config/buildnml
Comment (xylar) [inline components/mpas-ocean/cime_config/buildnml:813]: I think this one analysis member (`highFrequencyOutput`) probably should only have `landIceFreshwaterFluxTotal` and not the constituents.  We want to keep it pretty light weight.

**Comment by xylar** [comment]
@cbegeman and @jonbob, could you each have a look at this and let @irenavankova know if there is anything you would like to have happen before this moves over to E3SM?  If you would like to wait for my testing and MPAS-Analysis updates, that's fine.

**Comment by darincomeau** [comment]
Since I'm knee deep in the ocean conservation analysis member, does this work relate to/impact the existing variable `accumulatedLandIceFrazilFlux` at all?

**Comment by irenavankova** [comment]
> Since I'm knee deep in the ocean conservation analysis member, does this work relate to/impact the existing variable `accumulatedLandIceFrazilFlux` at all?

It doesn't affect it. It just keeps track of the nonaccumulated version of that variable.

**Comment by xylar** [comment]
It is only indirectly related to `accumulatedLandIceFrazilFlux`.  That variable is the sum over the whole simulation of the frazil flux, masked to under land ice.  `frazilIceFreshwaterFlux` is for the full ocean, not just under land ice and it is the instantaneous value (so basically the time derivative of `accumulatedLandIceFrazilFlux`) that would be passed to MALI as a freshwater flux in coupled simulations along with the melting/freezing at the interface.

File: components/mpas-ocean/cime_config/buildnml
Comment (cbegeman) [reply components/mpas-ocean/cime_config/buildnml:813]: I agree

File: components/mpas-ocean/src/shared/mpas_ocn_surface_land_ice_fluxes.F
Comment (cbegeman) [inline components/mpas-ocean/src/shared/mpas_ocn_surface_land_ice_fluxes.F:660]:
```suggestion
          do iCell = 1, nCells
             if (landIceFloatingMask(iCell) == 0) cycle
```
You might consider including this line as above to save computations. I haven't done a performance comparison though to see if it actually saves time if you only have a few FLOPs per cell.

File: components/mpas-ocean/src/shared/mpas_ocn_surface_land_ice_fluxes.F
Comment (xylar) [reply components/mpas-ocean/src/shared/mpas_ocn_surface_land_ice_fluxes.F:661]: There won't be frazil under grounded ice so it probably won't matter but your suggestion is more correct.

Comment (irenavankova) [reply components/mpas-ocean/src/shared/mpas_ocn_surface_land_ice_fluxes.F:661]: So I think it is a good idea to keep it consistent with the mask used to calculate accumulatedLandIceFrazilMass, at the moment the mask there is landIceMask:

https://github.com/E3SM-Project/E3SM/blob/dcc0ffb6db72edd3ce1ada559f4c2755d8dce15b/components/mpas-ocean/src/shared/mpas_ocn_frazil_forcing.F#L619

So do you suggest to change that to be modulated to landIceFloatingFraction also?

Comment (cbegeman) [reply components/mpas-ocean/src/shared/mpas_ocn_surface_land_ice_fluxes.F:661]: Thanks for pointing me to that code. I see that we shouldn't change to `landIceFloatingFraction` without making other changes as well. Since those changes could be non-BFB, I don't think we should tackle them right now (i.e., don't change anything here).

I think that if we were to change to `landIceFloatingFraction` we would want to account for the fact that open ocean frazil forms at a different salinity than that under land ice. I don't want to make any implicit assumptions that we only have fractional ice cells at the grounding line (i.e., that frazil would form in the non-floating fraction).

@xylar I was actually more concerned with some future scenario where we have fractional ice cells at the calving front (maybe when we have dynamic calving).

**Review by cbegeman — APPROVED** [review]
Approving based on visual inspection.

**Comment by cbegeman** [comment]
Thanks @irenavankova! This will be great to have!

File: components/mpas-ocean/src/shared/mpas_ocn_surface_land_ice_fluxes.F
Comment (xylar) [reply components/mpas-ocean/src/shared/mpas_ocn_surface_land_ice_fluxes.F:661]: Ah, sorry, it was late and I was tired. @irenavankova and I already discussed this.  We don't want to use the `landIceFloatingFraciton` in this case.  That handles the difference between floating and grounded ice but doesn't really handle half floating and half open ocean.  The frazil flux that is in such a cell would be lost, since it also doesn't go to the sea-ice model.  So I believe the mask is correct as it is.

Comment (xylar) [reply components/mpas-ocean/src/shared/mpas_ocn_surface_land_ice_fluxes.F:661]: I mistakenly thought the discussion was about `landIceMask` vs `landIceFloatingMask`.

File: components/mpas-ocean/src/shared/mpas_ocn_surface_land_ice_fluxes.F
Comment (xylar) [reply components/mpas-ocean/src/shared/mpas_ocn_surface_land_ice_fluxes.F:660]: I think that wouldn't make much difference in the grand scheme of things because most cores would just wait for the few that own land ice to finish.  But it doesn't hurt...

**Comment by xylar** [comment]
Analysis from my 5-year B-case run is here:
https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.xylar/analysis/20240226.CRYO1850.ne30pg2_r05_IcoswISC30E3r5.include_frazil.chrysalis/yrs_5-5/ocean/index.html

I was a little surprised that we get very little frazil in that case but it could be that conditions are just too warm or that 5 years isn't long enough.

**Comment by jonbob** [comment]
I think this is ready to move over to E3SM. Before we can merge it, we'll have to run the scripts to make the bld files consistent with Registry -- if only to pick up on the newer field descriptions

**Comment by irenavankova** [comment]
Thanks, @jonbob, I will move it over to E3SM now then. I am not sure which scripts you are referring to - is that something to do on my end?

**Pull request closed without merging**
